### PR TITLE
RFC: alert/noalert and config keyword extensions

### PR DESCRIPTION
**Pull request by victorjulien**
Other than the `alert` addition from #10154, this adds a way to disable stream reassembly completely and app-layer parsing (for TCP only atm).

E.g.
```
config smb any any -> any any (config:stream disable, type reassembly, scope flow; sid:1;)
config tcp any any -> any 445 (config:stream disable, type reassembly, scope flow; sid:2; alert;)
config tcp any any -> any 80 (flow:to_server; config:app-layer disable, type parser, scope flow; sid:3; alert;)
alert tcp any any -> any 80 (flow:to_server; content:"HTTP/1.0"; sid:4;)
```
Sid 1 and 2 both disable all stream reassembly for that flow.

Sid 3 only disables app parsers for on port 80. Content inspection still works (sid 4).

Would like some feedback on how to express things, and on what toggles to add.

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 181 | 192 | 106.08% | 

Pipeline 17500

File: src/detect-engine-alert.c [hunk 1]
```diff
@@ -407,22 +404,30 @@ void PacketAlertFinalize(DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx
...
             p->alerts.suppressed++;
         } else if (p->alerts.cnt < packet_alert_max) {
             p->alerts.alerts[p->alerts.cnt] = *pa;
             SCLogDebug("Appending sid %" PRIu32 " alert to Packet::alerts at pos %u", s->id, i);
 
-            /* pass "alert" found, we're done */
-            if (pa->action & ACTION_PASS) {
+            /* pass-alert found, we're done. Alert is not logged. */
+            if ((pa->action & (ACTION_PASS | ACTION_ALERT)) == ACTION_PASS) {
+                SCLogDebug("sid:%u: is a pass rule, so break out of out loop", s->id);
```
Comment (jufajardini) [inline src/detect-engine-alert.c:418]: Not sure if typo or not -> "out of out loop"

**Review by jufajardini — COMMENTED** [review]
> Other than the alert addition from https://github.com/OISF/suricata/pull/10154, this adds a way to disable stream reassembly completely and app-layer parsing (for TCP only atm).

Shouldn't we document this, as well as the `alert` keyword? (in the same section as `noalert`, the latter should go, I think).

Apart from that, all alert-related work seems to look good, to me. Only two minor nit comments (inline).
I saw that we have a SV test with a `drop noalert` rule that works as expected (`bug-4663`), wondering if there are other cases to ensure we cover with tests.

File: src/detect-config.c [hunk 2]
```diff
@@ -214,8 +252,9 @@ static int DetectConfigSetup (DetectEngineCtx *de_ctx, Signature *s, const char
         goto error;
     }
 
-    if (!(strcmp(typeval, "tx") == 0 ||strcmp(typeval, "flow") == 0)) {
-        SCLogError("only 'tx' and 'flow' supported at this time");
+    if (!(strcmp(typeval, "tx") == 0 || strcmp(typeval, "reassembly") == 0 ||
+                strcmp(typeval, "parser") == 0)) {
+        SCLogError("only 'tx', 'parser' and 'reassembly' supported at this time: %s", typeval);
```
Comment (inashivb) [inline src/detect-config.c:257]: nit: only ... **types** supported?

File: src/detect-config.c [hunk 3]
```diff
@@ -241,31 +280,62 @@ static int DetectConfigSetup (DetectEngineCtx *de_ctx, Signature *s, const char
     }
 
     if (!(strcmp(scopeval, "tx") == 0 ||strcmp(scopeval, "flow") == 0)) {
-        SCLogError("only 'tx' and 'flow' supported at this time");
+        SCLogError("only 'tx' and 'flow' supported at this time: %s", scopeval);
```
Comment (inashivb) [inline src/detect-config.c:283]: nit: **scopes** supported?

File: src/util-config.h [hunk 4]
```diff
@@ -40,6 +42,9 @@ enum ConfigType {
     CONFIG_TYPE_FILE,       /* file logging */
     CONFIG_TYPE_PCAP,       /* pcap logging */
     CONFIG_TYPE_DROP,       /* drop logging */
+    CONFIG_TYPE_REASSEMBLY, /**< stream reassembly */
+    CONFIG_TYPE_APP_PARSER, /**< app parsing */
+    CONFIG_TYPE_APP_FILES,  /**< app files */
```
Comment (inashivb) [inline src/util-config.h:47]: 1. Lots of TBD types here. Should they be marked?
2. Should the scope be made a part of type like it is for `APP` types?

**Review by inashivb — COMMENTED** [review]
LGTM mostly. Some questions, mostly nit, inline.

**Comment by inashivb** [comment]
> feedback on how to express things

1. not sure about the hyphen in `app-layer` in rule lang.
2. For stream, other type that comes to mind: `memcap` [allow to disable memcap limits when say one is interested in getting full stream matching a certain criteria]
3. `logging` makes more sense as a type to me than a subsystem so we can allow more control about what part of the system we want to disable logging on. Like, say, we want to disable logging but only for stream engine or an applayer parser or a tx (this means that tx seems more suited as a subsystem)

What do you think?

File: src/detect-engine-analyzer.c [hunk 5]
```diff
@@ -1003,7 +1004,7 @@ void EngineAnalysisRules2(const DetectEngineCtx *de_ctx, const Signature *s)
     if (s->flags & SIG_FLAG_DP_ANY) {
         jb_append_string(ctx.js, "dp_any");
     }
-    if (s->flags & SIG_FLAG_NOALERT) {
+    if ((s->action & ACTION_ALERT) == 0) {
         jb_append_string(ctx.js, "noalert");
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-analyzer.c:1008]: I think it was a quick hack but we should really add the actions, which we don't do yet.

**Comment by victorjulien** [comment]
> > feedback on how to express things
> 
>     1. not sure about the hyphen in `app-layer` in rule lang.

What would you suggest as an alternative? `app_layer`?

> 
>     2. For stream, other type that comes to mind: `memcap` [allow to disable memcap limits when say one is interested in getting full stream matching a certain criteria]

While this would be nice it will also be quite some work. Currently memcap is not tracked/handled with per flow/session limits, so changing that would require quite a bit of work.

>     3. `logging` makes more sense as a type to me than a subsystem so we can allow more control about what part of the system we want to disable logging on. Like, say, we want to disable logging but only for stream engine or an applayer parser or a tx (this means that tx seems more suited as a subsystem)

I conceptualize subsys as "capture", "decode", "stream", "app", "detect" and "output".

In any case, `logging` is already in 7 (and 6 I think), as a "subsys". Can rework that of course.

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 17531

**Comment by catenacyber** [comment]
Is there a redmine ticket for this ?

**Comment by jufajardini** [comment]
> Is there a redmine ticket for this ?

For the pass-alert thing: https://redmine.openinfosecfoundation.org/issues/5466

**Comment by catenacyber** [comment]
Any update on this ?

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 17531

File: src/detect-engine-analyzer.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-analyzer.c:1008]: On further thought, I do think adding `noalert` for a `pass` makes sense. We don't alert if the rule matches.

**Comment by victorjulien** [comment]
Replaced by #11260.

**Pull request closed without merging**